man/vlogger.1
.TH vlogger 1 "MAY 2023" "0.3.3" "fiss man page"
.PP
.SH NAME
.PP
\fB\fCvlogger\fR - log messages to syslog or an arbitrary executable 
.PP
.PP
.SH SYNOPSIS
.PP
\fB\fCvlogger\fR [\fB\fC-isS\fR] [\fB\fC-f\fR file] [\fB\fC-p\fR pri] [\fB\fC-t\fR tag] [message ...] 
.PP
.PP
.SH DESCRIPTION
.PP
The vlogger utility writes messages to the system log or an arbitrary executable. 
.PP
.PP
If vlogger is executed as logger it will always use the system log and behave like the regular logger(1). 
.PP
.PP
Without message arguments vlogger reads messages from stdin or the file specified with the -f flag. If the /etc/vlogger executable exists vlogger executes it with tag, level and facility as arguments, replacing the vlogger process. 
.PP
.PP
If vlogger is executed as a log service for fiss(8) or another daemontools like supervision suite it uses the service name as default tag. As example if vlogger is linked to /var/service/foo/log/run it uses “foo” as tag and “daemon.notice” as pri. 
.PP
.PP
The options are as follows: 
.PP
\fB\fC-f file\fR Read lines from the specified file. This option cannot be combine message arguments. 
.PP
\fB\fC-i\fR Log the PID of the vlogger process. Only supported if syslog(3) is used. 
.PP
\fB\fC-p pri\fR The. pri can be facility.level or just facility. See FACILITIES, LEVELS or syslog(3). The default is “user.notice”. 
.PP
\fB\fC-S\fR Force vlogger to use syslog(3) even if /etc/vlogger exists. 
.PP
\fB\fC-s\fR Output the message to standard error, as well as syslog(3). Only supported if syslog(3) is used. 
.PP
\fB\fC-t tag\fR Defines the openlog(3) ident which is used as prefix for each log message or passed as first argument to /etc/vlogger. The default is the LOGNAME environment variable. 
.PP
\fB\fCmessage\fR Write the message to the system log. 
.PP
.SH FACILITIES
.PP
auth
.PP
authpriv
.PP
cron
.PP
daemon
.PP
ftp
.PP
kern can not be used from userspace replaced with daemon
.PP
lpr
.PP
mail
.PP
news
.PP
syslog
.PP
user
.PP
uucp
.PP
local[0-7]
.PP
security deprecated synonym for auth 
.PP
.PP
.SH LEVELS
.PP
emerg
.PP
alert
.PP
crit
.PP
err
.PP
warning
.PP
notice
.PP
info
.PP
debug
.PP
panic deprecated synonym for emerg
.PP
error deprecated synonym for err
.PP
warn deprecated synonym for warning
.PP
.PP
.PP
.SH FILES
\fB\fC/etc/vlogger\fR
An optional executable file that is used to handle the messages. It is executed with tag, level and facility as arguments and replaces the vlogger process. 
.PP
.SH EXIT STATUS
.PP
The vlogger utility exits 0 on success, and >0 if an error occurs. 
.PP
.PP
.SH SEE ALSO
.PP
logger(1), syslog(3), svlogd(8) 
.PP
.PP
.SH HISTORY
.PP
This program is a replacement for the logger utility provided by util-linux and forked from daemontools. 
.PP
.PP
.SH AUTHORS
.PP
Duncan Overbruck <mail@duncano.de> 
.PP
.PP
.SH LICENSE
.PP
vlogger is in the public domain. 
.PP
To the extent possible under law, the creator of this work has waived all copyright and related or neighboring rights to this work. 
.PP
http://creativecommons.org/publicdomain/zero/1.0/ 


src/exec/vlogger.c
#include "config.h"
#include "message.h"
#include "util.h"

#include <errno.h>
#include <libgen.h>
#include <limits.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <sys/syslog.h>
#include <unistd.h>

static char pwd[PATH_MAX];

typedef struct ident {
	const char* name;
	int         value;
} ident_t;

ident_t prioritynames[] = {
	{ "alert", LOG_ALERT },
	{ "crit", LOG_CRIT },
	{ "debug", LOG_DEBUG },
	{ "emerg", LOG_EMERG },
	{ "err", LOG_ERR },
	{ "error", LOG_ERR },
	{ "info", LOG_INFO },
	{ "notice", LOG_NOTICE },
	{ "panic", LOG_EMERG },
	{ "warn", LOG_WARNING },
	{ "warning", LOG_WARNING },
	{ 0, -1 }
};

ident_t facilitynames[] = {
	{ "auth", LOG_AUTH },
	{ "authpriv", LOG_AUTHPRIV },
	{ "cron", LOG_CRON },
	{ "daemon", LOG_DAEMON },
	{ "ftp", LOG_FTP },
	{ "kern", LOG_KERN },
	{ "lpr", LOG_LPR },
	{ "mail", LOG_MAIL },
	{ "news", LOG_NEWS },
	{ "security", LOG_AUTH },
	{ "syslog", LOG_SYSLOG },
	{ "user", LOG_USER },
	{ "uucp", LOG_UUCP },
	{ "local0", LOG_LOCAL0 },
	{ "local1", LOG_LOCAL1 },
	{ "local2", LOG_LOCAL2 },
	{ "local3", LOG_LOCAL3 },
	{ "local4", LOG_LOCAL4 },
	{ "local5", LOG_LOCAL5 },
	{ "local6", LOG_LOCAL6 },
	{ "local7", LOG_LOCAL7 },
	{ 0, -1 }
};

static void strpriority(char* facil_str, int* facility, int* level) {
	char*    prio_str = NULL;
	ident_t* ident;

	if ((prio_str = strchr(facil_str, '.'))) {
		*prio_str = '\0';
		prio_str++;
		for (ident = prioritynames; ident->name; ident++) {
			if (streq(ident->name, prio_str))
				*level = ident->value;
		}
	}
	if (*facil_str) {
		for (ident = facilitynames; ident->name; ident++) {
			if (streq(ident->name, facil_str))
				*facility = ident->value;
		}
	}
}

int main(int argc, char* argv[]) {
	char  buf[SV_VLOGGER_BUFFER];
	char *p, *e, *argv0;
	char* tag = NULL;
	int   c;
	bool  Sflag    = false;
	int   logflags = 0;
	int   facility = LOG_USER;
	int   level    = LOG_NOTICE;

	argv0 = *argv;

	if (streq(argv0, "./run")) {
		// if running as a service, update facility and tag
		p = getcwd(pwd, sizeof(pwd));
		if (p != NULL && *pwd == '/') {
			if (*(p = pwd + (strlen(pwd) - 1)) == '/')
				*p = '\0';
			if ((p = strrchr(pwd, '/')) && strncmp(p + 1, "log", 3) == 0 &&
			    (*p = '\0', (p = strrchr(pwd, '/'))) && (*(p + 1) != '\0')) {
				tag      = p + 1;
				facility = LOG_DAEMON;
				level    = LOG_NOTICE;
			}
		}
	} else if (streq(basename(argv0), "logger")) {
		/* behave just like logger(1) and only use syslog */
		Sflag = true;
	}

	while ((c = getopt(argc, argv, "f:ip:Sst:")) != -1)
		switch (c) {
			case 'f':
				if (freopen(optarg, "r", stdin) == NULL) {
					print_error("error: unable to reopen %s: %s\n", optarg);
					return 1;
				}
				break;
			case 'i':
				logflags |= LOG_PID;
				break;
			case 'p':
				strpriority(optarg, &facility, &level);
				break;
			case 'S':
				Sflag = true;
				break;
			case 's':
				logflags |= LOG_PERROR;
				break;
			case 't':
				tag = optarg;
				break;
			default:
				print_usage_exit(PROG_VLOGGER, 1);
		}
	argc -= optind;
	argv += optind;

	if (argc > 0)
		Sflag = true;

	if (!Sflag && access("/etc/vlogger", X_OK) != -1) {
		ident_t*    ident;
		const char *sfacility = "", *slevel = "";
		for (ident = prioritynames; ident->name; ident++) {
			if (ident->value == level)
				slevel = ident->name;
		}
		for (ident = facilitynames; ident->name; ident++) {
			if (ident->value == facility)
				sfacility = ident->name;
		}
		execl("/etc/vlogger", argv0, tag ? tag : "", slevel, sfacility, NULL);
		print_error("error: unable to exec /etc/vlogger: %s\n");
		exit(1);
	}

	openlog(tag ? tag : getlogin(), logflags, facility);

	if (argc > 0) {
		size_t len;
		p  = buf;
		*p = '\0';
		e  = buf + sizeof(buf) - 2;
		while (*argv) {
			len = strlen(*argv);
			if (p + len > e && p > buf) {
				syslog(level | facility, "%s", buf);
				p  = buf;
				*p = '\0';
			}
			if (len > sizeof(buf) - 1) {
				syslog(level | facility, "%s", *argv++);
			} else {
				if (p != buf) {
					*p++ = ' ';
					*p   = '\0';
				}
				strncat(p, *argv++, e - p);
				p += len;
			}
		}
		if (p != buf)
			syslog(level | facility, "%s", buf);
		return 0;
	}

	while (fgets(buf, sizeof(buf), stdin) != NULL)
		syslog(level | facility, "%s", buf);

	return 0;
}
